Add DokuWiki to mash

shukon opened the pull request:
Hello!
I'd like to add support for the established [DokuWiki](https://www.dokuwiki.org).
I've tested the setup for about 6 months, updates were smooth, I'd like to offer it to the community. I'll gladly handle any further modifications necessary. It's not really a big or complicated contribution, but it's a service more.
edit: [ansible-repo](https://github.com/shukon/ansible-role-dokuwiki)

spatterIight commented on templates/group_vars_mash_servers:
```diff
@@ -2020,6 +2025,46 @@ docker_registry_purger_gid: "{{ mash_playbook_gid }}"
+#                                                                      #
+########################################################################
+# /role-specific:dokuwiki
 
```
I believe the convention in this file is to have 3 new lines in-between sections, could you add them to bottom (here) and top (above)?

shukon replied:
Ah, sure. Done. I spotted some where it's different, so I probably didn't catch the pattern.

spatterIight commented:
I think this would be a really awesome addition, ever since Confluence went cloud only self-hosted Wiki solutions have been in demand

spatterIight commented on docs/supported-services.md:
```diff
@@ -20,6 +20,7 @@
 | [Docker Registry Browser](https://github.com/klausmeyer/docker-registry-browser) | Web Interface for the Docker Registry HTTP API V2 written in Ruby on Rails | [Link](services/docker-registry-browser.md) |
 | [Docker Registry Proxy](https://gitlab.com/etke.cc/docker-registry-proxy/) | Pass-through docker registry (distribution) proxy with metadata caching, docker-compatible errors, prometheus metrics, etc. | [Link](services/docker-registry-proxy.md) |
 | [Docker Registry Purger](https://github.com/devture/docker-registry-purger) | A small tool used for purging a private Docker Registry's old tags | [Link](services/docker-registry-purger.md) |
+| [Dokuwiki](https://dokuwiki.org) | A wiki engine with simple syntax that works without a database | [Link](services/dokuwiki.md) |
```
```suggestion
| [DokuWiki](https://dokuwiki.org/) | A lightweight, file-based wiki engine with intuitive syntax and no database requirements | [Link](services/dokuwiki.md) |
```

spatterIight commented on docs/services/dokuwiki.md:
```diff
@@ -0,0 +1,39 @@
+# Dokuwiki
```
```suggestion
# DokuWiki
```

spatterIight commented on docs/services/dokuwiki.md:
```diff
@@ -0,0 +1,39 @@
+# Dokuwiki
+
+[Dokuwiki](https://dokuwiki.org/) is a common wiki engine that does not require a database.
```
```suggestion
[DokuWiki](https://dokuwiki.org/) is a lightweight, file-based wiki engine with intuitive syntax and no database requirements.
``

spatterIight commented on templates/group_vars_mash_servers:
```diff
@@ -2020,6 +2025,46 @@ docker_registry_purger_gid: "{{ mash_playbook_gid }}"
+
+########################################################################
+#                                                                      #
+# /dokuwiki                                                              #
```
```suggestion
# /dokuwiki                                                            #
```

spatterIight commented on templates/group_vars_mash_servers:
```diff
@@ -2020,6 +2025,46 @@ docker_registry_purger_gid: "{{ mash_playbook_gid }}"
+# role-specific:dokuwiki
+########################################################################
+#                                                                      #
+# dokuwiki                                                               #
```
```suggestion
# dokuwiki                                                             #
```

shukon commented:
Thanks for the review and for catching my oversights. I actually only learned it's Doku**W**iki now, after > 5 years of working with it :sweat_smile:

spatterIight commented on docs/services/dokuwiki.md:
```diff
@@ -0,0 +1,40 @@
+# DokuWiki
+
+[DokuWiki](https://dokuwiki.org/) is a lightweight, file-based wiki engine with intuitive syntax and no database requirements.
+``
```
```suggestion

```

spatterIight commented:
In the role there are some variables defined in `defaults/main.yml` that are not actually used anywhere:

```
dokuwiki_container_image_force_pull
dokuwiki_container_http_host_bind_port
dokuwiki_container_labels_traefik_additional_request_headers
```

To use these variables I believe the following conventions apply:

`dokuwiki_container_image_force_pull`:

```yaml
- name: Ensure qBittorrent container image is pulled via community.docker.docker_image
  when: devture_systemd_docker_base_container_image_pull_method == 'ansible-module'
  community.docker.docker_image:
    name: "{{ qbittorrent_container_image }}"
    source: "{{ 'pull' if ansible_version.major > 2 or ansible_version.minor > 7 else omit }}"
    force_source: "{{ syncthing_container_image_force_pull if ansible_version.major > 2 or ansible_version.minor >= 8 else omit }}"
    force: "{{ omit if ansible_version.major > 2 or ansible_version.minor >= 8 else syncthing_container_image_force_pull }}"
  register: image_pull_result
  retries: "{{ devture_playbook_help_container_retries_count }}"
  delay: "{{ devture_playbook_help_container_retries_delay }}"
  until: image_pull_result is not failed
```


`dokuwiki_container_http_host_bind_port`:

In systemd service
```yaml
			{% if sonarr_container_http_bind_port %}
			-p {{ sonarr_container_http_bind_port }}:{{ sonarr_container_http_port }} \
			{% endif %}
```

In traefik labels:

```yaml
traefik.http.services.{{ sonarr_identifier }}.loadbalancer.server.port={{ sonarr_container_http_port }}
```

`dokuwiki_container_labels_traefik_additional_request_headers`:

not sure about this one actually, never seen it before and none of my roles have it. probably fine to just remove this unless an official MASH maintainer says otherwise.

shukon commented:
Mmh - do you think I can just remove them all? They are left-overs from the template-role I used (not sure which one it was exactly). Or are they relevant for PR-acceptance?
I could add the snippets you mentioned (thanks for that!), but I'm not sure I will find time to test them properly soon...

spantaleev commented:
Sure, you can remove all of these.

That said, it's probably a good idea to keep `dokuwiki_container_image_force_pull` and make use of it in your "pull image" command, as it's very useful for when people pin the container image to `:latest`.

----------

I've also noticed that your "pull image" command is needlessly complicated (using a `with_items` loop), [here](https://github.com/shukon/ansible-role-dokuwiki/blob/e8ea48b505921cb33ceb8aeeef88c9a33db030ed/tasks/install.yml#L19-L24).

Why not just use the template you see [above](https://github.com/mother-of-all-self-hosting/mash-playbook/pull/386#issuecomment-2695130022), coming from the qBittorrent role.

That one also contains retry logic - see `devture_playbook_help_container_retries_count` and `devture_playbook_help_container_retries_delay`. All of our other roles have this and it helps make them more resilient to temporary container registry failures.

-------

Speaking of the qBittorrent role ["pull image" example above](https://github.com/mother-of-all-self-hosting/mash-playbook/pull/386#issuecomment-2695130022), I've noticed that it refers to `syncthing_` variables instead of `qbittorrent_` variables. This is a copy/paste mistake that I've since fixed [here](https://github.com/mother-of-all-self-hosting/ansible-role-qbittorrent/commit/f7c28087baa0a1ae725ca86118123726fde7eaf1).

--------

If you care about IPv6, your role may also make its "create container network" task better by respecting the `devture_systemd_docker_base_ipv6_enabled` variable, [like this](https://github.com/mother-of-all-self-hosting/ansible-role-qbittorrent/blob/2856b7c6734e201ce4d7effa8628785e338ed53b/tasks/install.yml#L48-L54).

--------

Generally, it's better to copy/paste an existing role (preferably one made by the MASH team and more actively maintained) and to do replacements there, without re-inventing the wheel.

Throwing stuff away and doing it from scratch means:

- your role is more alien and doesn't integrate as well with the rest of the playbook (and its roles)
- your role doesn't support all the functionality that people may come to expect. For example:
  - all other roles implement force-pulling for container images when the container image is pointing to a `latest` tag.
  - many roles do retries when trying to pull container images
  - many roles have IPv6 support by respecting the `devture_systemd_docker_base_ipv6_enabled` variable
  - many roles respect the `devture_systemd_docker_base_docker_service_name` variable, instead of hardcoding `docker.service`. See an example [here](https://github.com/mother-of-all-self-hosting/ansible-role-qbittorrent/blob/2856b7c6734e201ce4d7effa8628785e338ed53b/defaults/main.yml#L52-L56)
  - etc.
  
You don't strictly need to do all of this (and not all roles do it yet), but the more you do, the less users of your role (via the MASH playbook) will be surprised.

shukon commented:
Thank you for the detailed and thoughtful feedback.

I did attempt to copy templates - but that was six months ago and maybe I took a non-maintained role. Nevermind, I took the qbittorrent role as a template now and rearranged things. Still, I removed `dokuwiki_container_http_host_bind_port`
and `dokuwiki_container_labels_traefik_additional_request_headers`, because I don't think the role needs them (?). Happy to integrate them if you think otherwise.

I did add IPv6 and pull-retries and changed the overall look to conform better with the playbook-standards.
